packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD2.hpp

```diff
@@ -0,0 +1,1323 @@
+      recoloringPerRound[0] = 0;
+      commPerRound[0] = comm_time;
+      compPerRound[0] = interior_time + conflict_detection;
+      conflictDetectionPerRound[0] = conflict_detection;
+      recoloringPerRound[0] = 0;
+      vertsPerRound[0] = 0;
+      //see if recoloring is necessary.
+      typename Kokkos::View<int*, device_type>::HostMirror colors_host;
+      typename Kokkos::View<lno_t*, device_type>::HostMirror verts_to_recolor_host;
+      typename Kokkos::View<lno_t*, device_type>::HostMirror verts_to_send_host;
+      typename Kokkos::View<int*, device_type>::HostMirror ghost_colors_host;
+      int serial_threshold = this->pl->template get<int>("serial_threshold",0);
+      
+      gno_t totalConflicts = 0;
+      gno_t localConflicts = recoloringSize(0);
+      Teuchos::reduceAll<int,gno_t>(*comm, Teuchos::REDUCE_SUM, 1, &localConflicts, &totalConflicts);
+      bool done = !totalConflicts;
+      if(comm->getSize()==1) done = true;
+      
+      //recolor until no conflicts are left
+      while(!done){
+	if(recoloringSize(0) < serial_threshold) break;
+        if(distributedRounds < 100){
+         vertsPerRound[distributedRounds] = recoloringSize(0);
+        }
+  
+        comm->barrier();
+        double recolor_temp = timer();
+        //recolor using KokkosKernels' coloring function 
+	if(verts_to_recolor_size(0) > 0){
+          this->colorInterior<execution_space, memory_space, memory_space>
+		  (femv_colors.size(), dist_adjs_dev, dist_offsets_dev, femv, verts_to_recolor_view, verts_to_recolor_size(0), false);
+	}
+        recoloringPerRound[distributedRounds] = timer() - recolor_temp;
+        recoloring_time += recoloringPerRound[distributedRounds];
+        comp_time += recoloringPerRound[distributedRounds];
+        compPerRound[distributedRounds] = recoloringPerRound[distributedRounds];
+        totalPerRound[distributedRounds] = recoloringPerRound[distributedRounds];
+        recoloringSize(0) = 0;
+	//verts_to_recolor_size(0) = 0;
+        Kokkos::parallel_for(rand.size() - n_local, KOKKOS_LAMBDA(const int& i){
+          femv_colors(i+n_local) = ghost_colors(i);
+        });
+        Kokkos::fence();
+        //communicate the new colors
+        comm->barrier();
+	std::cout<<comm->getRank()<<":verts_to_send_size(0) = "<<verts_to_send_size(0)<<"\n";
+        commPerRound[distributedRounds] = doOwnedToGhosts(mapWithCopies,n_local, verts_to_send_view, verts_to_send_size,femv_colors);
+        comm_time += commPerRound[distributedRounds];
+        totalPerRound[distributedRounds] += commPerRound[distributedRounds];
+	//verts_to_send_size(0) = 0;
+        Kokkos::parallel_for(rand.size() - n_local, KOKKOS_LAMBDA(const int& i){
+          ghost_colors(i) = femv_colors(i+n_local);
+        });
+        comm->barrier();
+
+        double detection_temp = timer();
+        Kokkos::parallel_reduce(boundary_verts_size, KOKKOS_LAMBDA(const uint64_t& i,gno_t& recoloring_size){
```

Comment:
This conflict detection code is repeated three times in the class:  near lines 774, here, and later on host.  Make it a private method, templated on execution and memory space, rather than repeating in three places.